File: modules/076-skills-management-system/tools/scripts/cmd/push.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"github.com/vibecoding/skills-cli/internal/config"
	"github.com/vibecoding/skills-cli/internal/gitops"
)

// RunPush handles the `skills push <skill-name> [--groups <g1> <g2> ...]` command.
func RunPush(args []string) {
	// Parse args manually to support: skills push <name> --groups g1 g2
	var skillName string
	var groups []string
	help := false
	collectingGroups := false

	for _, arg := range args {
		if arg == "--help" || arg == "-h" {
			help = true
			break
		}
		if arg == "--groups" {
			collectingGroups = true
			continue
		}
		if collectingGroups {
			if strings.HasPrefix(arg, "--") {
				collectingGroups = false
			} else {
				for _, g := range strings.Split(arg, ",") {
					g = strings.TrimSpace(g)
					if g != "" {
						groups = append(groups, g)
					}
				}
				continue
			}
		}
		if !strings.HasPrefix(arg, "--") && skillName == "" {
			skillName = arg
		}
	}

	if help {
		printPushHelp()
		return
	}

	if skillName == "" {
		fmt.Fprintln(os.Stderr, "Error: skill name is required")
		fmt.Fprintln(os.Stderr, "Usage: skills push <skill-name> [--groups <group1> <group2> ...]")
		os.Exit(1)
	}

	branchName := fmt.Sprintf("feature/%s-update", skillName)

	cfg, err := config.Load()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}

	repoDir := cfg.RepoPath()

	fmt.Printf("→ Creating branch %s ...\n", branchName)
	if err := gitops.CreateBranch(repoDir, branchName); err != nil {
		fmt.Fprintf(os.Stderr, "Error: failed to create branch: %v\n", err)
		fmt.Fprintf(os.Stderr, "Tip: if the branch already exists, delete it with:\n")
		fmt.Fprintf(os.Stderr, "     git -C instructions branch -D %s\n", branchName)
		os.Exit(1)
	}
	fmt.Println("  ✓ Branch created")

	fmt.Printf("→ Staging and committing changes in %s/ ...\n", skillName)
	if err := gitops.StageAndCommit(repoDir, skillName); err != nil {
		fmt.Fprintf(os.Stderr, "Error: commit failed: %v\n", err)
		fmt.Fprintln(os.Stderr, "Tip: make sure you have changes to commit in instructions/"+skillName+"/")
		_ = gitops.CheckoutBranch(repoDir, gitops.DefaultBranch(repoDir))
		os.Exit(1)
	}
	fmt.Println("  ✓ Changes committed")

	// Optional: add skill to group manifests
	if len(groups) > 0 {
		fmt.Printf("→ Adding skill to group manifest(s): %s ...\n", strings.Join(groups, ", "))
		manifestChanged := false
		for _, group := range groups {
			if addSkillToGroupManifest(repoDir, skillName, group) {
				manifestChanged = true
				fmt.Printf("  ✓ Added to %q\n", group)
			}
		}
		if manifestChanged {
			_ = gitops.AddToSparseCheckout(repoDir, ".manifest")
			if _, err := runGit(repoDir, "add", ".manifest/"); err == nil {
				commitMsg := fmt.Sprintf("feat(%s): add to groups %s", skillName, strings.Join(groups, ", "))
				if _, err := runGit(repoDir, "commit", "-m", commitMsg); err == nil {
					fmt.Println("  ✓ Manifest changes committed")
				} else {
					fmt.Fprintf(os.Stderr, "Warning: failed to commit manifest changes: %v\n", err)
				}
			}
		}
	}

	fmt.Printf("→ Pushing branch %s ...\n", branchName)
	if err := gitops.Push(repoDir, branchName); err != nil {
		fmt.Fprintf(os.Stderr, "Error: push failed: %v\n", err)
		_ = gitops.CheckoutBranch(repoDir, gitops.DefaultBranch(repoDir))
		os.Exit(1)
	}
	fmt.Println("  ✓ Branch pushed")

	fmt.Printf("\n✅ Skill %q pushed for review\n", skillName)
	fmt.Printf("   Branch: %s\n", branchName)
	if len(groups) > 0 {
		fmt.Printf("   Groups: %s\n", strings.Join(groups, ", "))
	}

	if remoteURL, err := gitops.GetRemoteURL(repoDir); err == nil {
		if prURL := buildPRURL(remoteURL, branchName); prURL != "" {
			fmt.Printf("   Create PR: %s\n", prURL)
		} else {
			fmt.Println("   (local repository - request a review from the skill owner)")
		}
	}
	fmt.Printf("\n⚠  Note: switched back to the main branch - skill %q may not be visible locally.\n", skillName)
	fmt.Println("   After the PR is merged, run `skills pull` to get it back.")
}

func printPushHelp() {
	fmt.Print(`Create a branch, commit local changes to a skill, and push for review.

Usage:
  skills push <skill-name> [--groups <group1> <group2> ...]

The command will:
  1. Create branch: feature/<skill-name>-update
  2. Stage all changes in instructions/<skill-name>/
  3. Commit with a conventional commit message
  4. (optional) Add skill to specified group manifests and commit manifest changes
  5. Push the branch to origin
  6. Print the Pull Request URL (for GitHub/GitLab remotes)

Flags:
  --groups  Add skill to specified group manifests (creates group file if not found)

Examples:
  skills push my-skill
  skills push my-skill --groups backend security
  skills push my-skill --groups backend,frontend

Note: when --groups is used, manifest changes are included in the same PR branch.
If a group manifest does not exist, it will be created with the skill as its first entry.
`)
}

type groupManifestFile struct {
	Skills     []string `json:"skills"`
	SubConfigs []string `json:"sub-configs"`
}

func addSkillToGroupManifest(repoDir, skillName, groupName string) bool {
	manifestDir := filepath.Join(repoDir, ".manifest")
	manifestFile := filepath.Join(manifestDir, groupName+".json")

	var m groupManifestFile

	if data, err := os.ReadFile(manifestFile); err == nil {
		if err := json.Unmarshal(data, &m); err != nil {
			m = groupManifestFile{Skills: []string{}, SubConfigs: []string{}}
		}
		for _, s := range m.Skills {
			if s == skillName {
				fmt.Printf("  ℹ Skill %q already in group %q\n", skillName, groupName)
				return false
			}
		}
	} else {
		// Create new group manifest
		if err := os.MkdirAll(manifestDir, 0755); err != nil {
			fmt.Fprintf(os.Stderr, "Error: failed to create .manifest/: %v\n", err)
			return false
		}
		m = groupManifestFile{Skills: []string{}, SubConfigs: []string{}}
		fmt.Printf("  → Creating new group manifest: %s.json\n", groupName)
	}

	m.Skills = append(m.Skills, skillName)
	sort.Strings(m.Skills)
	data, _ := json.MarshalIndent(m, "", "  ")
	if err := os.WriteFile(manifestFile, append(data, '\n'), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error: failed to write %s.json: %v\n", groupName, err)
		return false
	}
	return true
}

// runGit is a local helper that runs a raw git command (used for manifest staging).
func runGit(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	out, err := cmd.CombinedOutput()
	return strings.TrimSpace(string(out)), err
}

// buildPRURL converts a git remote URL into a PR creation URL for supported hosts.
func buildPRURL(remoteURL, branchName string) string {
	remoteURL = strings.TrimSuffix(strings.TrimSpace(remoteURL), ".git")
	switch {
	case strings.HasPrefix(remoteURL, "https://github.com/"):
		return remoteURL + "/compare/" + branchName
	case strings.HasPrefix(remoteURL, "git@github.com:"):
		return "https://github.com/" + strings.TrimPrefix(remoteURL, "git@github.com:") + "/compare/" + branchName
	case strings.HasPrefix(remoteURL, "https://gitlab.com/"):
		return remoteURL + "/-/merge_requests/new?merge_request[source_branch]=" + branchName
	default:
		return ""
	}
}

File: modules/076-skills-management-system/tools2/src/commands/push.ts
import * as fs from 'fs';
import * as path from 'path';
import * as config from '../lib/config';
import * as gitops from '../lib/gitops';

function parseArgs(args: string[]): { skillName: string; groups: string[]; help: boolean } {
  if (args.includes('--help') || args.includes('-h')) {
    return { skillName: '', groups: [], help: true };
  }

  let skillName = '';
  const groups: string[] = [];
  let i = 0;
  let groupsFlag = false;

  while (i < args.length) {
    if (args[i] === '--groups' && i + 1 < args.length) {
      groupsFlag = true;
      i++;
      // Collect all following non-flag args as group names
      while (i < args.length && !args[i].startsWith('--')) {
        for (const g of args[i].split(/[,\s]+/)) {
          const t = g.trim();
          if (t) groups.push(t);
        }
        i++;
      }
      continue;
    } else if (!args[i].startsWith('--') && !skillName) {
      skillName = args[i];
    }
    i++;
  }

  return { skillName, groups, help: false };
}

function printPushHelp(): void {
  console.log(`Create a branch, commit local changes to a skill, and push for review.

Usage:
  skills push <skill-name> [--groups <group1> <group2> ...]

The command will:
  1. Create branch: feature/<skill-name>-update
  2. Stage all changes in instructions/<skill-name>/
  3. Commit with a conventional commit message
  4. (optional) Add skill to specified group manifests and commit manifest changes
  5. Push the branch to origin
  6. Print the Pull Request URL (for GitHub/GitLab remotes)

Flags:
  --groups  Add skill to specified group manifests (creates group file if not found)

Examples:
  skills push my-skill
  skills push my-skill --groups backend security
  skills push my-skill --groups backend,frontend

Note: when --groups is used, manifest changes are included in the same PR branch.
If a group manifest does not exist, it will be created with the skill as its first entry.
`);
}

function addSkillToGroupManifest(repoDir: string, skillName: string, groupName: string): boolean {
  const manifestDir = path.join(repoDir, '.manifest');
  const manifestFile = path.join(manifestDir, `${groupName}.json`);

  let manifest: { skills: string[]; 'sub-configs'?: string[] };

  if (fs.existsSync(manifestFile)) {
    try {
      manifest = JSON.parse(fs.readFileSync(manifestFile, 'utf8'));
    } catch {
      manifest = { skills: [], 'sub-configs': [] };
    }
    if (!manifest.skills) manifest.skills = [];
    if (manifest.skills.includes(skillName)) {
      console.log(`  ℹ Skill "${skillName}" already in group "${groupName}"`);
      return false;
    }
  } else {
    // Create new group manifest
    if (!fs.existsSync(manifestDir)) {
      fs.mkdirSync(manifestDir, { recursive: true });
    }
    manifest = { skills: [], 'sub-configs': [] };
    console.log(`  → Creating new group manifest: ${groupName}.json`);
  }

  manifest.skills.push(skillName);
  manifest.skills.sort();
  fs.writeFileSync(manifestFile, JSON.stringify(manifest, null, 2) + '\n', 'utf8');
  return true;
}

export function runPush(args: string[]): void {
  const { skillName, groups, help } = parseArgs(args);

  if (help) {
    printPushHelp();
    return;
  }

  if (!skillName) {
    console.error('Error: skill name is required');
    console.error('Usage: skills push <skill-name> [--groups <group1> <group2> ...]');
    process.exit(1);
  }

  const branchName = `feature/${skillName}-update`;

  try {
    config.load();
  } catch (err) {
    console.error(String(err));
    process.exit(1);
  }
  const repoDir = config.REPO_SUB_DIR;

  console.log(`→ Creating branch ${branchName} ...`);
  try {
    gitops.createBranch(repoDir, branchName);
  } catch (err) {
    console.error(`Error: failed to create branch: ${err}`);
    console.error(`Tip: if the branch already exists, delete it with:`);
    console.error(`     git -C instructions branch -D ${branchName}`);
    process.exit(1);
  }
  console.log('  ✓ Branch created');

  console.log(`→ Staging and committing changes in ${skillName}/ ...`);
  try {
    gitops.stageAndCommit(repoDir, skillName);
  } catch (err) {
    console.error(`Error: commit failed: ${err}`);
    console.error(`Tip: make sure you have changes to commit in instructions/${skillName}/`);
    // Return to default branch before exiting so workspace is not left on a feature branch
    try { gitops.checkoutBranch(repoDir, gitops.defaultBranch(repoDir)); } catch { /* ignore */ }
    process.exit(1);
  }
  console.log('  ✓ Changes committed');

  // Optional: add skill to group manifests
  if (groups.length > 0) {
    console.log(`→ Adding skill to group manifest(s): ${groups.join(', ')} ...`);
    let manifestChanged = false;
    for (const group of groups) {
      if (addSkillToGroupManifest(repoDir, skillName, group)) {
        manifestChanged = true;
        console.log(`  ✓ Added to "${group}"`);
      }
    }
    if (manifestChanged) {
      try {
        gitops.addToSparseCheckout(repoDir, '.manifest');
        // Stage and commit manifest changes using execFileSync directly
        const { execFileSync } = require('child_process');
        execFileSync('git', ['add', '.manifest/'], { cwd: repoDir, encoding: 'utf8' });
        execFileSync('git', ['commit', '-m', `feat(${skillName}): add to groups ${groups.join(', ')}`], { cwd: repoDir, encoding: 'utf8' });
        console.log('  ✓ Manifest changes committed');
      } catch (err) {
        console.error(`Warning: failed to commit manifest changes: ${err}`);
      }
    }
  }

  console.log(`→ Pushing branch ${branchName} ...`);
  try {
    gitops.push(repoDir, branchName);
  } catch (err) {
    console.error(`Error: push failed: ${err}`);
    try { gitops.checkoutBranch(repoDir, gitops.defaultBranch(repoDir)); } catch { /* ignore */ }
    process.exit(1);
  }
  console.log('  ✓ Branch pushed');

  // Build PR URL hint for GitHub/GitLab
  let prUrl = '';
  try {
    const remoteURL = gitops.getRemoteURL(repoDir);
    const normalized = remoteURL
      .replace(/\.git$/, '')
      .replace(/^git@github\.com:/, 'https://github.com/')
      .replace(/^ssh:\/\/git@github\.com\//, 'https://github.com/')
      .replace(/^git@gitlab\.com:/, 'https://gitlab.com/')
      .replace(/^ssh:\/\/git@gitlab\.com\//, 'https://gitlab.com/');
    if (normalized.includes('github.com')) {
      prUrl = `${normalized}/compare/${branchName}?expand=1`;
    } else if (normalized.includes('gitlab.com')) {
      prUrl = `${normalized}/-/merge_requests/new?merge_request%5Bsource_branch%5D=${branchName}`;
    }
  } catch { /* ignore */ }

  console.log(`\n✅ Skill "${skillName}" pushed for review`);
  console.log(`   Branch: ${branchName}`);
  if (groups.length > 0) {
    console.log(`   Groups: ${groups.join(', ')}`);
  }
  if (prUrl) {
    console.log(`   Create PR: ${prUrl}`);
  } else {
    console.log('   (local repository - request a review from the skill owner)');
  }
  console.log(`\n⚠  Note: switched back to the main branch - skill "${skillName}" may not be visible locally.`);
  console.log('   After the PR is merged, run `skills pull` to get it back.');
}
